pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/GateImplementationsLM.hpp

```diff
@@ -672,25 +576,23 @@ class GateImplementationsLM : public PauliGenerator<GateImplementationsLM> {
             all_wires.insert(all_wires.begin(), wires.begin(), wires.end());
             all_wires.insert(all_wires.begin() + wires.size(),
                              controlled_wires.begin(), controlled_wires.end());
-            const auto revs =
-                reverseWires(num_qubits, all_wires, controlled_values);
+            const auto revs = reverseWires(num_qubits, all_wires, {});
             const auto &rev_wires = revs.first;
-            const auto &rev_wire_shifts = revs.second;
             const std::vector<std::size_t> parity =
                 Pennylane::Util::revWireParity(rev_wires);
-
+            std::vector<std::size_t> indices =
+                Gates::generateBitPatterns(wires, num_qubits);
+            Gates::controlBitPatterns(indices, num_qubits, controlled_wires,
+                                      controlled_values);
+            const std::size_t i0 = indices[0];
+            const std::size_t i1 = indices[1];
             PL_LOOP_PARALLEL(1)
             for (std::size_t k = 0; k < exp2(num_qubits - nw_tot); k++) {
-                std::size_t i0 = (k & parity[0]);
-                for (std::size_t i = 1; i < parity.size(); i++) {
-                    i0 |= ((k << i) & parity[i]);
-                }
-                for (std::size_t i = 0; i < n_contr; i++) {
-                    i0 &= ~(one << rev_wires[i]);
-                    i0 |= rev_wire_shifts[i];
+                std::size_t offset{0U};
```

Comment:
Could we put the code block below into a functor?

Reply:
Or function? Do you mean having `parity_2_offset` function?